### DISCUSSION TITLE
MCMC for lens model  DESIJ2158+0257

### DISCUSSION THREAD

**Discussion by Rajit13** [AstroBridge/BDLensing#69] [Q&A] [answered]
I am currently using a different cell to run MCMC for my lens model. which looks like this
![image](https://github.com/ajshajib/BDLensing/assets/81336262/571ddee7-80f1-42b1-aba2-fb89b1841111)

I've several questions
1.  I've used `20 runs` for now which took me $2^h15^m$. What should be the optimum `runs`?
2. Shouldn't I see some corner plots after running MCMC? So what is the code and `sampler` for MCMC for my particular case?
3. As I have a separate `kwargs_result_mcmc` so should I incorporate this here
```
model_plot = ModelPlot(
    multi_band_list, kwargs_model, kwargs_result, kwargs_result_mcmc, 
    arrow_size=0.02,
    cmap_string="cubehelix",
    image_likelihood_mask_list=kwargs_likelihood["image_likelihood_mask_list"]
)
```

An example code file where MCMC has been run and plotted would be beneficial for us. I checked the whole `lenstronomy` repository where I found only two instances  ([here](https://github.com/sibirrer/strong_lensing_lectures/blob/0da949eedd4a4b97d81507b84a2db309ed292ff7/Lectures/einstein_ring_III.ipynb#L336) and [here](https://github.com/lenstronomy/lenstronomy-tutorials/blob/db146dba098a98743c98989910498a4fae1f2edb/Notebooks/TimeDelayCosmography/cosmographic_uncertainty_estimation_and_forecasting_with_kinematics.ipynb#L421)) of using MCMC in a different context.

**Comment by ajshajib** [comment]
Pinging @mirsazzathossain, @rafee-adnan, and @aaimtiaz, if they could help with this. 👆

**Comment by aaimtiaz** [comment, marked as answer]
1. It depends on individuals' system and how good your initials are. I guess 500-800 would do the job.
2. 
`
    
    sampler_type, samples_mcmc, param_mcmc, dist_mcmc  = chain_list_mcmc

    param_class = fitting_seq.param_class

    print("number of non-linear parameters in the MCMC process: ",
          len(param_mcmc))
    print("parameters in order: ", param_mcmc)
    print("number of evaluations in the MCMC process: ",
          np.shape(samples_mcmc)[0])
    n_sample = len(samples_mcmc)
    print(n_sample)
    samples_mcmc_cut = samples_mcmc[int(n_sample*1/2.):]
    if not samples_mcmc == []:
        n, num_param = np.shape(samples_mcmc_cut)
        plot = corner.corner(samples_mcmc_cut[:,:], labels=param_mcmc[:],
                             show_titles=True)`

This is the sample code we used to see the corner plot while we were modeling the moke lenses. This could help you out.

**Reply by Rajit13** [reply]
Thanks for the code, it runs but it's not showing any results @aaimtiaz! I used -- 
```
import corner

if len(chain_list_mcmc) > 1:
    sampler_type, samples_mcmc, param_mcmc, dist_mcmc = chain_list_mcmc[1]

    param_class = fitting_seq_mcmc.kwargs_params['kwargs_lens']['num_param']

    print("number of non-linear parameters in the MCMC process: ", len(param_mcmc))
    print("parameters in order: ", param_mcmc)
    print("number of evaluations in the MCMC process: ", np.shape(samples_mcmc)[0])
    n_sample = len(samples_mcmc)
    print(n_sample)
    samples_mcmc_cut = samples_mcmc[int(n_sample*1/2.):]
    if not samples_mcmc_cut == []:
        n, num_param = np.shape(samples_mcmc_cut)
        plot = corner.corner(samples_mcmc_cut[:, :], labels=param_mcmc[:], show_titles=True)
```

**Comment by Rajit13** [comment]
I also printed out these
```
Number of parameters in param_mcmc: 22
Parameters in order: ['theta_E_lens0', 'gamma_lens0', 'e1_lens0', 'e2_lens0', 'center_x_lens0', 'center_y_lens0', 'gamma1_lens1', 'gamma2_lens1', 'R_sersic_source_light0', 'e1_source_light0', 'e2_source_light0', 'center_x_source_light0', 'center_y_source_light0', 'beta_source_light1', 'R_sersic_lens_light0', 'e1_lens_light0', 'e2_lens_light0', 'center_x_lens_light0', 'center_y_lens_light0', 'R_sersic_lens_light1', 'e1_lens_light1', 'e2_lens_light1']

```

@ajshajib can you check if the parameters are correct?

**Reply by ajshajib** [reply]
Great, looks fine to me.

**Comment by Rajit13** [comment]
I was able to plot the corners with some additional coding. So using `'n_burn': 200, 'n_run': 300, 'n_walkers': 80` I got the below plot

![corner2](https://github.com/ajshajib/BDLensing/assets/81336262/02fec9f2-168e-4ad6-8ef9-ba002207b046)

**Reply by ajshajib** [reply]
I usually set `n_burn = 0`, the required number of burnin steps can be thrown away after the MCMC is run. I am more interested to see the trace plot before the corner plot.

**Reply by Rajit13** [reply]
Tried my best to enhance the plot (but failed it seems)
![trace_plots(1)](https://github.com/ajshajib/BDLensing/assets/81336262/1d71b371-aca0-45ce-92b2-85bdbc29bcaa)

**Comment by Rajit13** [comment]
I have got a better corner plot after `'n_burn': 200, 'n_run': 400, 'n_walkers': 80`
![corner](https://github.com/ajshajib/BDLensing/assets/81336262/198ae54c-b917-490d-b56c-73c530bcef0e)

**Reply by ajshajib** [reply]
Have you checked for MCMC convergence?

**Reply by Rajit13** [reply]
No, I'm not sure how to check it. I plotted the trace but the graph has very poor graphics. Should I use the `thin`  variable as we did for MCMC previously?

**Reply by ajshajib** [reply]
You don't need to use `thin` for now, but you can ask me what it's for during a meeting. If the plotting of the MCMC trace is not working, using `thin` wouldn't fix it.
